### PR TITLE
IDEMPIERE-5147  Improve Query.first() methods - performance

### PR DESCRIPTION
**Pull request by igorpojzl**
Jira Ticket: https://idempiere.atlassian.net/browse/IDEMPIERE-5147

Question: Do i need to move 
`String sql = buildSQL(null, true);`
in TRY block? 
buildSQL can throw IllegalStateException. Which can cause that pageSize is not set to original value.

### CODE REVIEW

**Comment by CarlosRuiz-globalqss** [comment]
> Jira Ticket: https://idempiere.atlassian.net/browse/IDEMPIERE-5147
> 
> Question: Do i need to move `String sql = buildSQL(null, true);` in TRY block? buildSQL can throw IllegalStateException. Which can cause that pageSize is not set to original value.

Yes, please

**Comment by igorpojzl** [comment]
Commited.

**Review by CarlosRuiz-globalqss — CHANGES_REQUESTED** [review]
Hi @igorpojzl

I tested this against oracle and the impact on the database is too high, because the poor implementation of the paging for oracle.

Can you please integrate the following patch:
[pr1124_PR.txt](https://github.com/idempiere/idempiere/files/7847290/pr1124_PR.txt)

I added changes to the oracle paging - and the same works perfect for postgresql.

Also added a unit test case for this paging, and it works fine.

File: org.idempiere.test/src/org/idempiere/test/base/QueryTest.java [hunk 1]
```diff
@@ -230,7 +230,42 @@ public void testFirstIdOnly() throws Exception
 			.firstIdOnly();
 		});
 	}
-	
+
+	@Test
+	public void testPaging() {
+		System.setProperty("org.idempiere.db.debug", "true");
```
Comment (CarlosRuiz-globalqss) [inline org.idempiere.test/src/org/idempiere/test/base/QueryTest.java:236]: Sorry @igorpojzl  - can you please delete this line - I left that by mistake

**Review by CarlosRuiz-globalqss — APPROVED** [review]

**Pull request merged**